File: py_payment_api/app/services/store/payment/one_time/one_time_payment_handler.py
from ..abstract import PaymentHandler
from ..abstract import BasePaymentData
from typing import TypeVar, Generic, List, Optional, Any
from ..abstract.item_collection_service import ItemCollectionService
from enum import StrEnum
from .one_time_payment_data import OneTimePaymentData
from ....store.enum import ItemType

ITEM_CATEGORY = TypeVar("ITEM_CATEGORY", bound=StrEnum)
T = TypeVar("T", bound=OneTimePaymentData[Any])
class OneTimePaymentHandler(Generic[ITEM_CATEGORY, T], PaymentHandler[ITEM_CATEGORY, T]):

    def __init__(self, category: ITEM_CATEGORY, logger: Optional[ItemCollectionService[ITEM_CATEGORY]] = None) -> None:
        super().__init__(category, logger=logger)

    def onPayment(self, payment: T) -> None:
        if payment["status"] not in ("paid", "webhook_recieved", "sent_to_websocket", "sent_to_processor"):
            raise ValueError(f"Invalid payment status: {payment['status']}")
        print(f"Payment received for {payment['item_category']} item")
    

File: tsPaymentApi/src/services/store/payment/one_time/OneTimePaymentHandler.ts
import { PaymentHandler } from '../abstract/PaymentHandler';
import { ItemCollectionService } from '../abstract/ItemCollectionService';
import { OneTimePaymentData } from './OneTimePaymentData';

export class OneTimePaymentHandler<TCategory extends string> extends PaymentHandler<TCategory, OneTimePaymentData<TCategory>> {
    constructor(category: TCategory, logger?: ItemCollectionService<TCategory>) {
        super(category, logger);
    }

    public async onPayment(payment: OneTimePaymentData<TCategory>): Promise<void> {
        if (!['paid', 'webhook_recieved', 'sent_to_websocket', 'sent_to_processor'].includes(payment.status)) {
            throw new Error(`Invalid payment status: ${payment.status}`);
        }
        
        console.log(`Payment received for ${payment.itemCategory} item`);
    }
} 